# Part 1: Dataframe Construction
(for other parts of the assignment please scroll down and look for parts 2 and 3)

In [3]:
import pandas as pd

Let's load the Toronto neighborhood database as a table.

In [31]:
# The code was removed by Watson Studio for sharing.

In [32]:
Toronto_df.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned                        NaN
1         M2A      Not assigned                        NaN
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

Drop "Not assigned" values in column Borough.

In [33]:
Toronto_df=Toronto_df[Toronto_df.Borough != 'Not assigned'].reset_index(drop=True)
Toronto_df.head()

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

Check for empty cells or "Not assigned" values in Neighborhood column.

In [7]:
import numpy as np

print(len(Toronto_df[Toronto_df.Neighborhood == 'Not assigned']))
print(Toronto_df.Neighborhood.isnull().sum() )
print(len(Toronto_df[Toronto_df.Neighborhood == '']))

0
0
0


After droping "Not assigned" Borough there are no Neighborhoods which have no value or are valued "Not assigned".

Next let's check if there are duplicate values in Postal Code column.

In [8]:
print(len(set(Toronto_df['Postal Code'])) )
print(len(Toronto_df['Postal Code']) )

103
103


There are no duplicate values in Postal Code column. It seems that the case with "M5A" has been resolved already in Wikipedia page itself, as "M5A" already has two neighborhoods Regent Park, Harbourfront separated with a comma.
Let's  check if there is no duplicate neighbordhood names.

In [9]:
print(len(set(Toronto_df['Neighborhood'])) )

99


There are 4 duplicate neighborhood names. However, they have different postal codes, so essentially they are different neighborhoods. 
But to be able to distinguish between these neighborhoods, we have to create unique neighborhood name. We will achieve this by creating column which merges Postal code column and Neighborhood column.

In [34]:
Toronto_df['Neighborhood_PC']=Toronto_df['Postal Code']+' '+Toronto_df['Neighborhood']
Toronto_df.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

                                   Neighborhood_PC  
0                                    M3A Parkwoods  
1                             M4A Victoria Village  
2                    M5A Regent Park, Harbourfront  
3             M6A Lawrence Manor, Lawrence Heights  
4  M7A Queen's Park, Ontario Provincial Government

In [252]:
Toronto_df.shape

(103, 4)

Database has 103 rows.

# Part 2: Geographical coordinates
(for other parts of the assignment please scroll down and look for part 3, for part 1 scroll up)

Load the goegraphical coordinate database.

In [10]:

body = client_4ab6efedbd8545eeaf31c1dcdc1d649d.get_object(Bucket='applieddatasciencecapstone-donotdelete-pr-wiseierbro6z33',Key='Geospatial_Coordinates.csv')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

# If you are reading an Excel file into a pandas DataFrame, replace `read_csv` by `read_excel` in the next statement.
Geo_cord = pd.read_csv(body)
Geo_cord.head()


Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Add the coordinate info to the neighborhood databse.

In [35]:
Toronto_df = Toronto_df.merge(Geo_cord, on='Postal Code')
Toronto_df.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

                                   Neighborhood_PC   Latitude  Longitude  
0                                    M3A Parkwoods  43.753259 -79.329656  
1                             M4A Victoria Village  43.725882 -79.315572  
2                    M5A Regent Park, Harbourfront  43.654260 -79.360636  
3             M6A Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4  M7A Queen's Park, Ontario Provincial Government  43.662301 -79.389494

In [12]:
Toronto_df.shape

(103, 5)

# Part 3: Clustering neighborhoods.
(for others parts of the assignment please scroll up and look for parts 1 and 2)

Let's retrieve location coordinates of Toronto.

In [36]:
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

address = 'Toronto'

geolocator = Nominatim(user_agent="tor_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude



Install map rendering library.

In [14]:
!conda install -c conda-forge folium=0.5.0 --yes

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    ca-certificates-2020.4.5.1 |       hecc5488_0         146 KB  conda-forge
    openssl-1.1.1g             |       h516909a_0         2.1 MB  conda-forge
    certifi-2020.4.5.1         |   py36h9f0ad1d_0         151 KB  conda-forge
    folium-0.5.0               |             py_0          45 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    altair-4.1.0               |             py_1         614 KB  conda-forge
    branca-0.4.1               |             py_0          26 KB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    ------------------------------------------------------------
                       

Plot Toronto map with all neighborhoods with borough name, neighborhood name and postal code in the label.

In [37]:
import folium # map rendering library

map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, postal, borough, neighborhood in zip(Toronto_df['Latitude'], Toronto_df['Longitude'], Toronto_df['Postal Code'],Toronto_df['Borough'], Toronto_df['Neighborhood']):
    label = '{}, {}, {}'.format(neighborhood, borough, postal)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Load Foursqaure credentials.

In [43]:
CLIENT_ID = 'UNOJDCQH4GLAWEX422WINWFES5GAKQLLBSD135VKOTFPFZIL' #Foursquare ID
CLIENT_SECRET = '4CQFM1QAYYKXTM1MEJOIROCFFEHMUUWQFPZB01ADJNA1BJC0' #Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT=100 #maximum number of venues to retrieve

Create a function that allows to retrieve all venues in a 500m radius of a given location.

In [44]:
import requests
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood_PC', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Apply the given function to all Toronto neighborhoods and retrieve information on venues in the neighborhoods.

In [45]:
Toronto_venues = getNearbyVenues(names=Toronto_df['Neighborhood_PC'],
                                   latitudes=Toronto_df['Latitude'],
                                   longitudes=Toronto_df['Longitude']
                                  )

M3A Parkwoods
M4A Victoria Village
M5A Regent Park, Harbourfront
M6A Lawrence Manor, Lawrence Heights
M7A Queen's Park, Ontario Provincial Government
M9A Islington Avenue, Humber Valley Village
M1B Malvern, Rouge
M3B Don Mills
M4B Parkview Hill, Woodbine Gardens
M5B Garden District, Ryerson
M6B Glencairn
M9B West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
M1C Rouge Hill, Port Union, Highland Creek
M3C Don Mills
M4C Woodbine Heights
M5C St. James Town
M6C Humewood-Cedarvale
M9C Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
M1E Guildwood, Morningside, West Hill
M4E The Beaches
M5E Berczy Park
M6E Caledonia-Fairbanks
M1G Woburn
M4G Leaside
M5G Central Bay Street
M6G Christie
M1H Cedarbrae
M2H Hillcrest Village
M3H Bathurst Manor, Wilson Heights, Downsview North
M4H Thorncliffe Park
M5H Richmond, Adelaide, King
M6H Dufferin, Dovercourt Village
M1J Scarborough Village
M2J Fairview, Henry Farm, Oriole
M3J Northwood Park, York University
M4J East Toronto

In [46]:
print(Toronto_venues.shape)
print(len(Toronto_venues['Venue Category'].unique()))
Toronto_venues.head()

(2127, 7)
272


Neighborhood_PC  Neighborhood Latitude  Neighborhood Longitude  \
0         M3A Parkwoods              43.753259              -79.329656   
1         M3A Parkwoods              43.753259              -79.329656   
2         M3A Parkwoods              43.753259              -79.329656   
3  M4A Victoria Village              43.725882              -79.315572   
4  M4A Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude     Venue Category  
0         Brookbanks Park       43.751976       -79.332140               Park  
1          TTC stop #8380       43.752672       -79.326351           Bus Stop  
2           Variety Store       43.751974       -79.333114  Food & Drink Shop  
3  Victoria Village Arena       43.723481       -79.315635       Hockey Arena  
4             Tim Hortons       43.725517       -79.313103        Coffee Shop

In total in all neighborhoods 2127 venues were retrieved and 272 different venue categories were listed. <br>
Let's now see how many venues were retrieved by each Neighborhood.

In [51]:
Toronto_venues.groupby('Neighborhood_PC').count()

Neighborhood Latitude  \
Neighborhood_PC                                                             
M1B Malvern, Rouge                                                      1   
M1C Rouge Hill, Port Union, Highland Creek                              1   
M1E Guildwood, Morningside, West Hill                                   8   
M1G Woburn                                                              5   
M1H Cedarbrae                                                           9   
M1J Scarborough Village                                                 2   
M1K Kennedy Park, Ionview, East Birchmount Park                         5   
M1L Golden Mile, Clairlea, Oakridge                                     7   
M1M Cliffside, Cliffcrest, Scarborough Village ...                      2   
M1N Birch Cliff, Cliffside West                                         4   
M1P Dorset Park, Wexford Heights, Scarborough T...                      7   
M1R Wexford, Maryvale                                                   7   
M1S Agincourt                                                           4   
M1T Clarks Corners, Tam O'Shanter, Sullivan                            13   
M1V Milliken, Agincourt North, Steeles East, L'...                      2   
M1W Steeles West, L'Amoreaux West                                      12   
M2H Hillcrest Village                                                   5   
M2J Fairview, Henry Farm, Oriole                                       67   
M2K Bayview Village                                                     4   
M2L York Mills, Silver Hills                                            1   
M2N Willowdale, Willowdale East                                        33   
M2P York Mills West                                                     3   
M2R Willowdale, Willowdale West                                         5   
M3A Parkwoods                                                           3   
M3B Don Mills                                                           5   
M3C Don Mills                                                          22   
M3H Bathurst Manor, Wilson Heights, Downsview N...                     20   
M3J Northwood Park, York University                                     7   
M3K Downsview                                                           4   
M3L Downsview                                                           5   
...                                                                   ...   
M6B Glencairn                                                           4   
M6C Humewood-Cedarvale                                                  4   
M6E Caledonia-Fairbanks                                                 4   
M6G Christie                                                           16   
M6H Dufferin, Dovercourt Village                                       14   
M6J Little Portugal, Trinity                                           44   
M6K Brockton, Parkdale Village, Exhibition Place                       22   
M6L North Park, Maple Leaf Park, Upwood Park                            4   
M6M Del Ray, Mount Dennis, Keelsdale and Silver...                      5   
M6N Runnymede, The Junction North                                       4   
M6P High Park, The Junction South                                      23   
M6R Parkdale, Roncesvalles                                             14   
M6S Runnymede, Swansea                                                 38   
M7A Queen's Park, Ontario Provincial Government                        35   
M7R Canada Post Gateway Processing Centre                              12   
M7Y Business reply mail Processing Centre, Sout...                     18   
M8V New Toronto, Mimico South, Humber Bay Shores                       14   
M8W Alderwood, Long Branch                                              8   
M8X The Kingsway, Montgomery Road, Old Mill North                       2   
M8Y Old Mill South, King's Mill Park, Sunnylea,.

Now let's create a table with dummie variables showing what type of venue is listed in each row.

In [47]:
# one hot encoding
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Toronto_onehot['Neighborhood_PC'] = Toronto_venues['Neighborhood_PC'] 

# move neighborhood column to the first column
fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

Toronto_onehot.head()

Neighborhood_PC  Accessories Store  Afghan Restaurant  Airport  \
0         M3A Parkwoods                  0                  0        0   
1         M3A Parkwoods                  0                  0        0   
2         M3A Parkwoods                  0                  0        0   
3  M4A Victoria Village                  0                  0        0   
4  M4A Victoria Village                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Train Station  \
0                 0                    0  ...              0   
1                 0                    0  ...              0   
2                 0                    0  ...              0   
3                 0                    0  ...              0   
4                 0                    0  ...              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 273 columns]

Now group them and see frquency of each type by neighborhood.

In [48]:
Toronto_grouped = Toronto_onehot.groupby('Neighborhood_PC').mean().reset_index()
Toronto_grouped.head()

Neighborhood_PC  Accessories Store  \
0                          M1B Malvern, Rouge                0.0   
1  M1C Rouge Hill, Port Union, Highland Creek                0.0   
2       M1E Guildwood, Morningside, West Hill                0.0   
3                                  M1G Woburn                0.0   
4                               M1H Cedarbrae                0.0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                0.0      0.0                 0.0           0.0   
1                0.0      0.0                 0.0           0.0   
2                0.0      0.0                 0.0           0.0   
3                0.0      0.0                 0.0           0.0   
4                0.0      0.0                 0.0           0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0                  0.0   
1             0.0              0.0               0.0                  0.0   
2             0.0              0.0               0.0                  0.0   
3             0.0              0.0               0.0                  0.0   
4             0.0              0.0               0.0                  0.0   

   ...  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0  ...            0.0                            0.0               0.0   
1  ...            0.0                            0.0               0.0   
2  ...            0.0                            0.0               0.0   
3  ...            0.0                            0.0               0.0   
4  ...            0.0                            0.0               0.0   

   Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0          0.0                    0.0              0.0       0.0          0.0   
1          0.0                    0.0              0.0       0.0          0.0   
2          0.0                    0.0              0.0       0.0          0.0   
3          0.0                    0.0              0.0       0.0          0.0   
4          0.0                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 273 columns]

In [52]:
Toronto_grouped.shape

(100, 273)

Note that we have lost 3 rows. For 3 neighborhoods there were no venues to retrieve. We will deal with that leater. <br>
But now we will create a function that arranes the venues types in descening order by frequency.

In [54]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Get top 10 venues types for each neighborhood.

In [55]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood_PC']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
Toronto_venues_sorted = pd.DataFrame(columns=columns)
Toronto_venues_sorted['Neighborhood_PC'] = Toronto_grouped['Neighborhood_PC']

for ind in np.arange(Toronto_grouped.shape[0]):
    Toronto_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

Toronto_venues_sorted.head()

Neighborhood_PC 1st Most Common Venue  \
0                          M1B Malvern, Rouge  Fast Food Restaurant   
1  M1C Rouge Hill, Port Union, Highland Creek                   Bar   
2       M1E Guildwood, Morningside, West Hill     Electronics Store   
3                                  M1G Woburn           Coffee Shop   
4                               M1H Cedarbrae      Hakka Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0      Doner Restaurant          Dessert Shop    Dim Sum Restaurant   
1           Yoga Studio      Doner Restaurant    Dim Sum Restaurant   
2         Moving Target   Rental Car Location    Mexican Restaurant   
3    Mexican Restaurant     Korean Restaurant     Convenience Store   
4   Fried Chicken Joint                Lounge                  Bank   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                 Diner        Discount Store   Distribution Center   
1                 Diner        Discount Store   Distribution Center   
2        Breakfast Spot                  Bank          Intersection   
3           Yoga Studio      Doner Restaurant                 Diner   
4    Athletics & Sports  Caribbean Restaurant       Thai Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0               Dog Run            Donut Shop     Falafel Restaurant  
1               Dog Run            Donut Shop         Farmers Market  
2        Medical Center                 Diner         Discount Store  
3        Discount Store   Distribution Center                Dog Run  
4           Gas Station                Bakery                Dog Run

Next step is to cluster the neighborhoods based on frequency of venues in nieghborhoods.

In [58]:
from sklearn.cluster import KMeans 

# set number of clusters
kclusters = 5

Toronto_grouped_clustering = Toronto_grouped.drop('Neighborhood_PC', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([2, 0, 2, 2, 2, 4, 2, 2, 2, 2], dtype=int32)

Now added cluster labels to original table. We add 1 to cluster labels so to create space for the neighborhoods which did not have any venues retrieved from Foursqaure (they were not used in cluster analysis). 0 label later will be assigned to the "empty" neighborhoods. <br>
Then merge original neighborhood table with the venues table. Using left join we assure that the "empty" neighborhoods will be assigned 'NaN' values.

In [124]:
# add clustering labels
#Toronto_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_+1)

Toronto_merged = Toronto_df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
Toronto_merged = Toronto_merged.merge(Toronto_venues_sorted.set_index('Neighborhood_PC'), on='Neighborhood_PC',how='left')

Toronto_merged.head() # check the last columns!

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

                                   Neighborhood_PC   Latitude  Longitude  \
0                                    M3A Parkwoods  43.753259 -79.329656   
1                             M4A Victoria Village  43.725882 -79.315572   
2                    M5A Regent Park, Harbourfront  43.654260 -79.360636   
3             M6A Lawrence Manor, Lawrence Heights  43.718518 -79.464763   
4  M7A Queen's Park, Ontario Provincial Government  43.662301 -79.389494   

   Cluster Labels 1st Most Common Venue   2nd Most Common Venue  \
0             4.0                  Park       Food & Drink Shop   
1             3.0          Hockey Arena   Portuguese Restaurant   
2             3.0           Coffee Shop                    Park   
3             3.0        Clothing Store  Furniture / Home Store   
4             3.0           Coffee Shop        Sushi Restaurant   

  3rd Most Common Venue        4th Most Common Venue 5th Most Common Venue  \
0              Bus Stop  Eastern European Restaurant   Dumpling Restaurant   
1           Coffee Shop                  Pizza Place      Department Store   
2                   Pub                       Bakery                  Café   
3              Boutique           Miscellaneous Shop           Event Space   
4              Creperie                          Bar              Beer Bar   

  6th Most Common Venue 7th Most Common Venue  8th Most Common Venue  \
0             Drugstore            Donut Shop       Doner Restaurant   
1          Dessert Shop    Dim Sum Restaurant                  Diner   
2        Breakfast Spot               Theater            Yoga Studio   
3     Accessories Store           Coffee Shop  Vietnamese Restaurant   
4         Smoothie Shop        Sandwich Place           Burger Joint   

  9th Most Common Venue 10th Most Common Venue  
0           Event Space      Electronics Store  
1        Discount Store    Distribution Center  
2     Electronics Store         Ice Cream Shop  
3          Concert Hall     Dim Sum Restaurant  
4         Burrito Place                   Café

Assign 0s to "empty" neighborhoods.

In [126]:
where_are_NaNs = np.isnan(Toronto_merged['Cluster Labels'])
Toronto_merged.loc[where_are_NaNs,'Cluster Labels']=0

In [128]:
Toronto_merged=Toronto_merged.astype({'Cluster Labels':'int32'})

Plot clustered neighborhoods and assign specific color to each neighborhood.

In [131]:
import matplotlib.cm as cm
import matplotlib.colors as colors

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters+1)
ys = [i + x + (i*x)**2 for i in range(kclusters+1)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], Toronto_merged['Neighborhood'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Let's examine the clusters.

##### Cluster 1

In [134]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0, Toronto_merged.columns[[1,2] + list(range(6, Toronto_merged.shape[1]))]]

Borough                                       Neighborhood  \
11    Etobicoke  West Deane Park, Princess Gardens, Martin Grov...   
52   North York                            Willowdale, Newtonbrook   
95  Scarborough                                        Upper Rouge   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11               0                   NaN                   NaN   
52               0                   NaN                   NaN   
95               0                   NaN                   NaN   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
11                   NaN                   NaN                   NaN   
52                   NaN                   NaN                   NaN   
95                   NaN                   NaN                   NaN   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11                   NaN                   NaN                   NaN   
52                   NaN                   NaN                   NaN   
95                   NaN                   NaN                   NaN   

   9th Most Common Venue 10th Most Common Venue  
11                   NaN                    NaN  
52                   NaN                    NaN  
95                   NaN                    NaN

These neighborhoods did not have any venues listed in Fouraquare. We call in a simple way "Empty neighborhoods".

##### Cluster 2

In [135]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1, Toronto_merged.columns[[1,2] + list(range(6, Toronto_merged.shape[1]))]]

Borough                            Neighborhood  Cluster Labels  \
12  Scarborough  Rouge Hill, Port Union, Highland Creek               1   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
12                   Bar           Yoga Studio      Doner Restaurant   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
12    Dim Sum Restaurant                 Diner        Discount Store   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
12   Distribution Center               Dog Run            Donut Shop   

   10th Most Common Venue  
12         Farmers Market

The smallest cluster, it lists only one neighborhood. It seems that is defined by the number 1 venue - Bar, since other venues are quite common in other clusters too. We name it "Outgoing neighborhood".

##### Cluster 3

In [136]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 2, Toronto_merged.columns[[1,2] + list(range(6, Toronto_merged.shape[1]))]]

Borough                                       Neighborhood  \
57   North York                                   Humberlea, Emery   
101   Etobicoke  Old Mill South, King's Mill Park, Sunnylea, Hu...   

     Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
57                2        Baseball Field           Yoga Studio   
101               2        Baseball Field                  Pool   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
57     Dim Sum Restaurant                 Diner        Discount Store   
101           Yoga Studio               Dog Run          Dessert Shop   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
57    Distribution Center               Dog Run      Doner Restaurant   
101    Dim Sum Restaurant                 Diner        Discount Store   

    9th Most Common Venue 10th Most Common Venue  
57             Donut Shop         Farmers Market  
101   Distribution Center       Doner Restaurant

This is the second smallest cluster. It has Baseball field on the top, next is Yoga Studio. One of neighborhoods also have a pool. It will be called "Sporty Neighborhood" cluster.

##### Cluster 4

In [137]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 3, Toronto_merged.columns[[1,2] + list(range(6, Toronto_merged.shape[1]))]]

Borough                                       Neighborhood  \
1          North York                                   Victoria Village   
2    Downtown Toronto                          Regent Park, Harbourfront   
3          North York                   Lawrence Manor, Lawrence Heights   
4    Downtown Toronto        Queen's Park, Ontario Provincial Government   
6         Scarborough                                     Malvern, Rouge   
7          North York                                          Don Mills   
8           East York                    Parkview Hill, Woodbine Gardens   
9    Downtown Toronto                           Garden District, Ryerson   
10         North York                                          Glencairn   
13         North York                                          Don Mills   
14          East York                                   Woodbine Heights   
15   Downtown Toronto                                     St. James Town   
16               York                                 Humewood-Cedarvale   
17          Etobicoke  Eringate, Bloordale Gardens, Old Burnhamthorpe...   
18        Scarborough                  Guildwood, Morningside, West Hill   
19       East Toronto                                        The Beaches   
20   Downtown Toronto                                        Berczy Park   
22        Scarborough                                             Woburn   
23          East York                                            Leaside   
24   Downtown Toronto                                 Central Bay Street   
25   Downtown Toronto                                           Christie   
26        Scarborough                                          Cedarbrae   
27         North York                                  Hillcrest Village   
28         North York    Bathurst Manor, Wilson Heights, Downsview North   
29          East York                                   Thorncliffe Park   
30   Downtown Toronto                           Richmond, Adelaide, King   
31       West Toronto                       Dufferin, Dovercourt Village   
33         North York                       Fairview, Henry Farm, Oriole   
34         North York                    Northwood Park, York University   
36   Downtown Toronto  Harbourfront East, Union Station, Toronto Islands   
..                ...                                                ...   
67    Central Toronto                                   Davisville North   
68    Central Toronto    Forest Hill North & West, Forest Hill Road Park   
69       West Toronto                      High Park, The Junction South   
70          Etobicoke                                          Westmount   
71        Scarborough                                  Wexford, Maryvale   
72         North York                        Willowdale, Willowdale West   
73    Central Toronto                  North Toronto West, Lawrence Park   
74    Central Toronto                The Annex, North Midtown, Yorkville   
75       West Toronto                             Parkdale, Roncesvalles   
76        Mississauga              Canada Post Gateway Processing Centre   
78        Scarborough                                          Agincourt   
79    Central Toronto                                         Davisville   
80   Downtown Toronto                     University of Toronto, Harbord   
81       West Toronto                                 Runnymede, Swansea   
82        Scarborough            Clarks Corners, Tam O'Shanter, Sullivan   
83    Central Toronto                        Moore Park, Summerhill East   
84   Downtown Toronto          Kensington Market, Chinatown, Grange Park   
86    Central Toronto  Summerhill West, Rathnelly, South Hill, Forest...   
87   Downtown Toronto  CN Tower, King and Spadina, Railway Lands, Har...   
88          Etobicoke       New Toronto, Mimico South, Humber Bay Shores   
89          Etobicoke  South Steeles, Silverstone, Humbergate,

This cluster by far is the largest with 82 members. In the most of neighborhoods one of the most common venues is coffee shop. Other venues frequently in the top include Gym, Grocery store, Clothing store. In general neighborhoods in this cluster have wide variety of venues. We may call this cluster "Standard cosy neighborhood".

##### Cluster 5

In [138]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 4, Toronto_merged.columns[[1,2] + list(range(6, Toronto_merged.shape[1]))]]

Borough                                       Neighborhood  \
0         North York                                          Parkwoods   
21              York                                Caledonia-Fairbanks   
35         East York      East Toronto, Broadview North (Old East York)   
45        North York                           York Mills, Silver Hills   
61   Central Toronto                                      Lawrence Park   
64              York                                             Weston   
66        North York                                    York Mills West   
85       Scarborough  Milliken, Agincourt North, Steeles East, L'Amo...   
91  Downtown Toronto                                           Rosedale   
98         Etobicoke      The Kingsway, Montgomery Road, Old Mill North   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0                4                  Park     Food & Drink Shop   
21               4                  Park                  Pool   
35               4                  Park         Metro Station   
45               4                  Park           Yoga Studio   
61               4                  Park              Bus Line   
64               4                  Park     Convenience Store   
66               4     Convenience Store                  Park   
85               4                  Park            Playground   
91               4                  Park                 Trail   
98               4                  Park                 River   

         3rd Most Common Venue        4th Most Common Venue  \
0                     Bus Stop  Eastern European Restaurant   
21               Women's Store             Greek Restaurant   
35           Convenience Store            Electronics Store   
45            Doner Restaurant                 Dessert Shop   
61                 Swim School                      Dog Run   
64            Doner Restaurant                 Dessert Shop   
66  Construction & Landscaping                   Donut Shop   
85                     Dog Run             Department Store   
91                  Playground             Department Store   
98          College Rec Center              College Stadium   

          5th Most Common Venue        6th Most Common Venue  \
0           Dumpling Restaurant                    Drugstore   
21                  Curling Ice  Eastern European Restaurant   
35  Eastern European Restaurant          Dumpling Restaurant   
45           Dim Sum Restaurant                        Diner   
61           Dim Sum Restaurant                        Diner   
64           Dim Sum Restaurant                        Diner   
66           Dim Sum Restaurant                        Diner   
85                 Dessert Shop           Dim Sum Restaurant   
91                 Dessert Shop           Dim Sum Restaurant   
98            Electronics Store  Eastern European Restaurant   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0             Donut Shop      Doner Restaurant           Event Space   
21   Dumpling Restaurant             Drugstore            Donut Shop   
35             Drugstore            Donut Shop      Department Store   
45        Discount Store   Distribution Center               Dog Run   
61        Discount Store   Distribution Center      Doner Restaurant   
64        Discount Store   Distribution Center               Dog Run   
66        Discount Store   Distribution Center               Dog Run   
85                 Diner        Discount Store   Distribution Center   
91                 Diner        Discount Store   Distribution Center   
98   Dumpling Restaurant             Drugstore            Donut Shop   

   10th Most Common Venue  
0       Electronics Store  
21       Doner Restaurant  
35       Doner Restaurant  
45             Donut Shop  
61       Department Store  
64            Yoga Studio  
66       Doner Restaurant  
85       Doner Restaurant  
91              

This is second largest cluster with 10 entries. It's clear venue leader is Park. The most of neighborhoods also include Diner, Dim Sum Restaurant, Distribution Center, Discount store. The cluster could be called "The green dining neighborhood".

##### Cluster 6

In [139]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 5, Toronto_merged.columns[[1,2] + list(range(6, Toronto_merged.shape[1]))]]

Borough                                       Neighborhood  \
5     Etobicoke            Islington Avenue, Humber Valley Village   
32  Scarborough                                Scarborough Village   
50   North York                                      Humber Summit   
63         York                      Runnymede, The Junction North   
77    Etobicoke  Kingsview Village, St. Phillips, Martin Grove ...   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
5                5           Pizza Place           Yoga Studio   
32               5           Pizza Place            Playground   
50               5          Home Service           Pizza Place   
63               5               Brewery           Pizza Place   
77               5     Mobile Phone Shop              Bus Line   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
5       Doner Restaurant          Dessert Shop    Dim Sum Restaurant   
32           Yoga Studio               Dog Run      Department Store   
50           Yoga Studio          Dessert Shop    Dim Sum Restaurant   
63              Bus Line     Convenience Store      Doner Restaurant   
77           Pizza Place        Sandwich Place           Yoga Studio   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
5                  Diner        Discount Store   Distribution Center   
32          Dessert Shop    Dim Sum Restaurant                 Diner   
50                 Diner        Discount Store   Distribution Center   
63    Dim Sum Restaurant                 Diner        Discount Store   
77        Discount Store          Dessert Shop    Dim Sum Restaurant   

   9th Most Common Venue 10th Most Common Venue  
5                Dog Run             Donut Shop  
32        Discount Store    Distribution Center  
50               Dog Run       Doner Restaurant  
63   Distribution Center                Dog Run  
77                 Diner    Distribution Center

Here the top venues are Pizza Place, Yoga Studio. We may call it "Relaxed pizza lovers' neighborhood"